iocore/net/quic/QUICLossDetector.cc

```diff
@@ -0,0 +1,356 @@
+  }
+  // Find all newly acked packets.
+  for (auto acked_packet_number : this->_determine_newly_acked_packets(*ack_frame)) {
+    this->_on_packet_acked(acked_packet_number);
+  }
+
+  this->_detect_lost_packets(ack_frame->largest_acknowledged());
+  Debug("quic_loss_detector", "Unacked handshake pkt %u, retransmittable pkt %u", this->_handshake_outstanding,
+        this->_retransmittable_outstanding);
+  this->_set_loss_detection_alarm();
+}
+
+void
+QUICLossDetector::_on_packet_acked(QUICPacketNumber acked_packet_number)
+{
+  SCOPED_MUTEX_LOCK(lock, this->mutex, this_ethread());
+  Debug("quic_loss_detector", "Packet number %" PRIu64 " has been acked", acked_packet_number);
+  this->_largest_acked_packet = acked_packet_number;
+  // If a packet sent prior to RTO was acked, then the RTO
+  // was spurious.  Otherwise, inform congestion control.
+  if (this->_rto_count > 0 && acked_packet_number > this->_largest_sent_before_rto) {
+    // TODO cc->on_retransmission_timeout_verified();
+  }
+  this->_handshake_count = 0;
+  this->_tlp_count       = 0;
+  this->_rto_count       = 0;
+  this->_decrement_packet_count(acked_packet_number);
+  this->_sent_packets.erase(acked_packet_number);
+}
+
+void
+QUICLossDetector::_decrement_packet_count(QUICPacketNumber packet_number)
+{
+  SCOPED_MUTEX_LOCK(lock, this->mutex, this_ethread());
+  auto ite = this->_sent_packets.find(packet_number);
+  if (ite != this->_sent_packets.end()) {
+    if (ite->second->handshake) {
+      --this->_handshake_outstanding;
+    }
+    if (ite->second->retransmittable) {
+      --this->_retransmittable_outstanding;
+    }
+  }
+}
+
+void
+QUICLossDetector::_on_loss_detection_alarm()
+{
+  if (this->_handshake_outstanding) {
+    // Handshake retransmission alarm.
+    this->_retransmit_handshake_packets();
+    this->_handshake_count++;
+  } else if (this->_loss_time != 0) {
+    // Early retransmit or Time Loss Detection
+    this->_detect_lost_packets(this->_largest_acked_packet);
+  } else if (this->_tlp_count < this->_MAX_TLPS) {
+    // Tail Loss Probe.
+    // eventProcessor.schedule_imm(this->_handler, ET_CALL, QUIC_EVENT_RETRANSMIT_ONE_PACKET);
+    this->_tlp_count++;
+  } else {
+    // RTO.
+    if (this->_rto_count == 0) {
+      this->_largest_sent_before_rto = this->_largest_sent_packet;
+    }
+    // eventProcessor.schedule_imm(this->_handler, ET_CALL, QUIC_EVENT_RETRANSMIT_TWO_PACKET);
+    this->_rto_count++;
+  }
+  Debug("quic_loss_detector", "Unacked handshake pkt %u, retransmittable pkt %u", this->_handshake_outstanding,
+        this->_retransmittable_outstanding);
+  this->_set_loss_detection_alarm();
+}
+
+void
+QUICLossDetector::_update_rtt(uint32_t latest_rtt)
+{
+  // Based on {{RFC6298}}.
+  if (this->_smoothed_rtt == 0) {
+    this->_smoothed_rtt = latest_rtt;
+    this->_rttvar       = latest_rtt / 2;
+  } else {
+    this->_rttvar       = 3 / 4 * this->_rttvar + 1 / 4 * (this->_smoothed_rtt - latest_rtt);
```

Comment:
Do we need to use absolutely value. `this->_rttvar       = 3 / 4 * this->_rttvar + 1 / 4 * ABS(this->_smoothed_rtt - latest_rtt);`
`
   (2.3) When a subsequent RTT measurement R' is made, a host MUST set

            RTTVAR <- (1 - beta) * RTTVAR + beta * |SRTT - R'|
            SRTT <- (1 - alpha) * SRTT + alpha * R'

`

Reply:
Most of code in QUICLossDetector came from pseudo code on the specification. I'm not sure whether we need to use the values.
https://tools.ietf.org/html/draft-ietf-quic-recovery-04#section-3.2.5

Reply:
https://github.com/quicwg/base-drafts/issues/753